### PR TITLE
TA88175: cisco_vrf_af:pt:Add new puppet type for cisco_vrf_af for Nexus and XR

### PR DESCRIPTION
**Pull request by HelenHZhang**
Adding route_policy_export, route_policy_import, route_target_export_stitching, route_target_import_stitching.
All beaker tests passed on eXR and N9kv.

### CODE REVIEW

File: examples/cisco/demo_vrf.pp [hunk 1]
```diff
@@ -86,11 +96,15 @@
...
+    route_policy_export           => 'abc',
+    route_policy_import           => 'abc',
     route_target_import           => $rt_import,
     route_target_export           => $rt_export,
     route_target_both_auto        => $rt_both,
     route_target_import_evpn      => $rt_import_evpn,
+    route_target_import_stitching => $rt_import_stitching,
     route_target_export_evpn      => $rt_export_evpn,
+    route_target_export_stitching => $rt_import_stitching,
     route_target_both_auto_evpn   => $rt_both_evpn,
```
Comment (HelenHZhang) [inline examples/cisco/demo_vrf.pp]: fixed the typo. should be $rt_exporting_stitching

File: examples/cisco/demo_vrf.pp [hunk 2]
```diff
@@ -16,9 +16,19 @@
 
 class ciscopuppet::cisco::demo_vrf {
 
+  $rt_export_stitching = $operatingsystem ? {
+    'ios_xr' => ['22:13 stitching', '24:15 stitching'],
```
Comment (chrisvanheuveln) [inline examples/cisco/demo_vrf.pp]: Basic question: Why does the manifest have to specify `stitching` for this property? I mean, the property itself defines this as a stitching command so what is gained by specifying it in the manifest? It doesn't seem to be an optional keyword.

File: lib/puppet/type/cisco_vrf_af.rb [hunk 3]
```diff
@@ -192,6 +226,29 @@ def insync?(is)
...
 
+  newproperty(:route_target_import_stitching, array_matching: :all) do
+    desc '(stitching only) Set the route-target import stitching communities. '\
+         'Valid values are an Array, a space-separated String of extended '\
+         "communities, or the keyword 'default'."
+
+    match_error = 'must be specified in AS:nn or IPv4:nn notation'
+    validate do |community|
+      community.split.each do |value|
+        fail "Confederation peer value '#{value}' #{match_error}" unless
```
Comment (chrisvanheuveln) [inline lib/puppet/type/cisco_vrf_af.rb]: `Confederation peer` ?

File: lib/puppet/type/cisco_vrf_af.rb [hunk 4]
```diff
@@ -131,6 +145,26 @@ def name
 
   ensurable
 
+  newproperty(:route_policy_export) do
+    desc 'Set route-policy(ios xr) or map(nexus) export name. Valid value '\
+         'is string.'
```
Comment (chrisvanheuveln) [inline lib/puppet/type/cisco_vrf_af.rb]: `/'is string./"is string or keyword 'default'."/`

File: tests/beaker_tests/cisco_vrf_af/test_vrf_af.rb [hunk 5]
```diff
@@ -92,7 +98,14 @@
 def unsupported_properties(tests, id)
   unprops = []
   if operating_system == 'nexus'
-    unprops = tests[id][:manifest_props] unless /n(3|9)k/.match(platform)
+    if /n(3|9)k/.match(platform)
```
Comment (chrisvanheuveln) [inline tests/beaker_tests/cisco_vrf_af/test_vrf_af.rb]: Do we support stitching on n(5|6|7)k?

File: examples/cisco/demo_vrf.pp (hunk 2 above)
Comment (HelenHZhang) [inline examples/cisco/demo_vrf.pp]: You are correct. I'll correct this.

File: lib/puppet/type/cisco_vrf_af.rb (hunk 4 above)
Comment (HelenHZhang) [inline lib/puppet/type/cisco_vrf_af.rb]: correct it.

File: tests/beaker_tests/cisco_vrf_af/test_vrf_af.rb (hunk 5 above)
Comment (HelenHZhang) [inline tests/beaker_tests/cisco_vrf_af/test_vrf_af.rb]: I only tested this on eXR and N9kv. I don't know the answer on other platforms.

Comment (chrisvanheuveln) [inline tests/beaker_tests/cisco_vrf_af/test_vrf_af.rb]: Then you cannot commit this change.

File: lib/puppet/type/cisco_vrf_af.rb (hunk 3 above)
Comment (HelenHZhang) [inline lib/puppet/type/cisco_vrf_af.rb]: I actually did not find any documentation on this. From everything I tried, it works as route_target_import, except they are the stitching RT.

Comment (chrisvanheuveln) [inline lib/puppet/type/cisco_vrf_af.rb]: Sorry I wasn't clear -- this doesn't have anything to do with confederations so the message should not say "Confederation peer...", it should say "Route-target stitching..."

Comment (HelenHZhang) [inline lib/puppet/type/cisco_vrf_af.rb]: corrected it. Thanks for catching it.

File: README.md [hunk 6]
```diff
@@ -2658,14 +2658,18 @@ Manages Cisco Virtual Routing and Forwarding (VRF) Address-Family configuration.
...
-| route_target_export_evpn    | Only supported on N3k and N9k        |
-| route_target_import         | Only supported on N3k, N9k, and XR   |
-| route_target_import_evpn    | Only supported on N3k and N9k        |
+| Property                      | Caveat Description                   |
+|-------------------------------|--------------------------------------|
+| route_policy_export           | Only supported on N9kv and XR        |
+| route_policy_import           | Only supported on N9kv and XR        |
+| route_target_both_auto        | Only supported on N3k and N9k        |
+| route_target_both_auto_evpn   | Only supported on N3k and N9k        |
+| route_target_export           | Only supported on N3k, N9k, and XR   |
```
Comment (glennmatthews) [inline README.md]: Please use 'IOS XR' rather than 'XR' in documentation

File: lib/puppet/type/cisco_vrf_af.rb [hunk 7]
```diff
@@ -39,6 +40,19 @@
       route_target_import          => ['5:6'],
       route_target_import_evpn     => ['7:7'],
     }
+
+    (xr)
```
Comment (glennmatthews) [inline lib/puppet/type/cisco_vrf_af.rb]: ios_xr or IOS XR please

File: README.md [hunk 8]
```diff
@@ -2658,14 +2658,18 @@ Manages Cisco Virtual Routing and Forwarding (VRF) Address-Family configuration.
...
+| route_target_export_evpn      | Only supported on N3k and N9k        |
+| route_target_export_stitching | Only supported on XR                 |
+| route_target_import           | Only supported on N3k, N9k, and XR   |
+| route_target_import_evpn      | Only supported on N3k and N9k        |
+| route_target_import_stitching | Only supported on XR                 |
 
 #### Parameters
```
Comment (glennmatthews) [inline README.md]: Please add documentation of the four new parameters.

File: README.md [hunk 9]
```diff
@@ -2658,14 +2658,18 @@ Manages Cisco Virtual Routing and Forwarding (VRF) Address-Family configuration.
...
-| route_target_both_auto      | Only supported on N3k and N9k        |
-| route_target_both_auto_evpn | Only supported on N3k and N9k        |
-| route_target_export         | Only supported on N3k, N9k, and XR   |
-| route_target_export_evpn    | Only supported on N3k and N9k        |
-| route_target_import         | Only supported on N3k, N9k, and XR   |
-| route_target_import_evpn    | Only supported on N3k and N9k        |
+| Property                      | Caveat Description                   |
+|-------------------------------|--------------------------------------|
+| route_policy_export           | Only supported on N9kv and XR        |
+| route_policy_import           | Only supported on N9kv and XR        |
```
Comment (glennmatthews) [inline README.md]: What's keeping us from supporting these two properties on other Nexus platforms?

File: README.md (hunk 6 above)
Comment (HelenHZhang) [inline README.md]: corrected.

File: lib/puppet/type/cisco_vrf_af.rb (hunk 7 above)
Comment (HelenHZhang) [inline lib/puppet/type/cisco_vrf_af.rb]: ios_xr

File: README.md (hunk 8 above)
Comment (HelenHZhang) [inline README.md]: added documentation for these.

File: README.md (hunk 9 above)
Comment (HelenHZhang) [inline README.md]: worked with Chris, and resolved all issues.

**Comment by HelenHZhang** [comment]
resolved all comments, please re-review it.

**Comment by chrisvanheuveln** [comment]
:+1:

**Pull request merged**